### PR TITLE
refactor: register tree items

### PR DESCRIPTION
**Pull request by SamuelAlev**

### CODE REVIEW

**Review by GoranRibic — APPROVED** [review]

File: src/components/Tree/Tree.tsx [hunk 1]
```diff
@@ -1,93 +1,258 @@
...
         return () => {
-            window.removeEventListener('keydown', downKeyHandler);
-            window.removeEventListener('keyup', upKeyHandler);
+            window.removeEventListener('keydown', keyDownHandler);
+            window.removeEventListener('keyup', keyUpHandler);
         };
-    }, []);
+    }, [keyDownHandler, keyUpHandler]);
+
+    useEffect(() => {
```
Comment (GoranRibic) [inline src/components/Tree/Tree.tsx:185]: Isn't this `useEffect` redundant?

File: src/components/Tree/TreeItem.tsx [hunk 2]
```diff
@@ -9,33 +19,66 @@ import { DraggableItem, DropZonePosition } from '@utilities/dnd';
...
+import { getAcceptTypes, getItemPositionInParent, getNextItemToFocus, getPreviousItemToFocus } from './helpers';
+import { cloneThroughFragments, flattenChildren } from './utils';
 
 const DRAGGING_OPACITY = 0.4;
 const DEFAULT_OPACITY = 1;
 
 export const TreeItem = ({
     id,
-    sort,
+    parentId,
```
Comment (GoranRibic) [inline src/components/Tree/TreeItem.tsx:33]: why does TreeItem have a `parentId` prop? Should that be inside treeState?

Comment (GoranRibic) [reply src/components/Tree/TreeItem.tsx:33]: What is level used for?

File: src/components/Tree/TreeItem.tsx [hunk 3]
```diff
@@ -44,50 +87,122 @@ export const TreeItem = ({
...
+                        onExpand(id, false);
+                    } else if (treeItemState?.parentId) {
+                        const parentId = treeItemState.parentId;
+                        const parentState = treeState.items.get(parentId);
+                        parentState?.domElement?.focus();
+                    }
+                    break;
+
+                case 'ArrowUp':
+                    event.preventDefault();
```
Comment (GoranRibic) [inline src/components/Tree/TreeItem.tsx:167]: event.preventDefault() can be put on top

File: src/components/Tree/helpers/index.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+/* (c) Copyright Frontify Ltd., all rights reserved. */
+
+export * from './getAcceptedTypes';
+export * from './getItemPositionInParent';
+export * from './getNextItemToFocus';
+export * from './getPreviousItemToFocus';
```
Comment (GoranRibic) [inline src/components/Tree/helpers/index.ts:6]: IMO these should be different files and some of them combined. 
getAcceptedTypes -> dnd.ts
getItemPositionInParent -> position.ts
getNextItemToFocus, getPreviousItemToFocus -> focus.ts

or something similar in nature.

File: src/components/Tree/types.ts [hunk 5]
```diff
@@ -40,6 +44,32 @@ export type TreeItemProps = {
...
+    childrenIds?: string[];
+    level: number;
+    domElement?: HTMLElement;
+};
+
+export type TreeState = {
+    items: Map<string, TreeItemState>;
+    selectedIds: Set<string>;
+    expandedIds: Set<string>;
+    selectionMode: 'single' | 'multiselect';
```
Comment (GoranRibic) [inline src/components/Tree/types.ts:64]: multi-select should be allowed or disallowed through property multi-select in the tree component.

File: src/components/Tree/TreeItem.tsx (hunk 2 above)
Comment (SamuelAlev) [reply src/components/Tree/TreeItem.tsx:33]: > why does TreeItem have a parentId prop? Should that be inside treeState?
Passed by its parent when rendering happens

> What is level used for?
[aria-level](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-level)

File: src/components/Tree/TreeItem.tsx (hunk 3 above)
Comment (SamuelAlev) [reply src/components/Tree/TreeItem.tsx:167]: It prevents tabbing

File: src/components/Tree/helpers/index.ts (hunk 4 above)
Comment (SamuelAlev) [reply src/components/Tree/helpers/index.ts:6]: Easier to find the methods IMO
What would be the benefit?

File: src/components/Tree/Tree.tsx (hunk 1 above)
Comment (SamuelAlev) [reply src/components/Tree/Tree.tsx:185]: Different use-case

**Pull request merged**